lighter/callbacks/logger.py

```diff
@@ -0,0 +1,375 @@
+        if self.tensorboard:
+            self.tensorboard.add_image(name, image, global_step=global_step)
+        if self.wandb:
+            self.wandb.log({name: OPTIONAL_IMPORTS["wandb"].Image(image)}, step=global_step)
+
+    def _on_batch_end(self, outputs: Dict, trainer: Trainer) -> None:
+        """Performs logging at the end of a batch/step. It logs the loss and metrics,
+        and, if specified how, the input, target, and pred data.
+
+        Args:
+            outputs (Dict): output dict from the model.
+            trainer (Trainer): Trainer, passed automatically by PyTorch Lightning.
+        """
+        if not trainer.sanity_checking:
+            mode = LIGHTNING_TO_LIGHTER_STAGE[trainer.state.stage]
+            # Accumulate the loss.
+            if mode in ["train", "val"]:
+                self.loss[mode] += outputs["loss"].item()
+            # Logging frequency.
+            if self.global_step_counter[mode] % trainer.log_every_n_steps == 0:
+                # Log. Done only on rank 0.
+                self._log(outputs, mode, global_step=self._get_global_step(trainer))
+            # Increment the step counters.
+            self.global_step_counter[mode] += 1
+            if mode in ["train", "val"]:
+                self.epoch_step_counter[mode] += 1
+
+    def _on_epoch_end(self, trainer: Trainer, pl_module: LighterSystem) -> None:
+        """Performs logging at the end of an epoch. It calculates the average
+        loss and metrics for the epoch and logs them. In distributed mode, it averages
+        the losses and metrics from all processes.
+
+        Args:
+            trainer (Trainer): Trainer, passed automatically by PyTorch Lightning.
+            pl_module (LighterSystem): LighterSystem, passed automatically by PyTorch Lightning.
+        """
+        if not trainer.sanity_checking:
+            mode = LIGHTNING_TO_LIGHTER_STAGE[trainer.state.stage]
+            outputs = {"loss": None, "metrics": None}
+
+            # Loss
+            if mode in ["train", "val"]:
+                # Get the accumulated loss.
+                loss = self.loss[mode]
+                # Reduce the loss to rank 0 and average it.
+                if dist.is_initialized():
+                    # Distributed communication works only tensors.
+                    loss = torch.tensor(loss).to(pl_module.device)
+                    # On rank 0, sum the losses from all ranks. Other ranks remain with the same loss as before.
+                    dist.reduce(loss, dst=0)
+                    # On rank 0, average the loss sum by dividing it with the number of processes.
+                    loss = loss.item() / dist.get_world_size() if dist.get_rank() == 0 else loss.item()
+                # Divide the accumulated loss by the number of steps in the epoch.
+                loss /= self.epoch_step_counter[mode]
+                outputs["loss"] = loss
+            
+            # Metrics
+            # Get the torchmetrics.
```

Comment:
I think its a good idea also to have other metrics supported. But maybe a separate issue for later?

Reply:
Metrics have to be `torchmetrics`:
https://github.com/project-lighter/lighter/blob/f7ac3d6260c43894a93fcb8f0d755bbed90241de/lighter/system.py#L70

 I wouldn't support any other format of metric. I think that accounting for all the possible ways of doing it would quickly create a mess here:
  https://github.com/project-lighter/lighter/blob/f7ac3d6260c43894a93fcb8f0d755bbed90241de/lighter/callbacks/logger.py#L242
  
 If a user wants a metric that's not available in `torchmetrics` they should implement it as a `torchmetrics.Metric` themself. It's pretty easy https://torchmetrics.readthedocs.io/en/stable/pages/implement.html.

However,  from a quick glance it appears that Monai metrics have a unified API too - we could write a metric wrapper that turns a Monai metric into a TorchMetric metric. It'd go to our `contrib` repo.

What do you guys think?

Reply:
That sounds perfect then! Since `torchmetrics` handle so many things internally, I agree that this is the best approach

Reply:
Ok, I agree. But then the logger should raise an Exception, if a metric is passed, that is not a `torchmetrics.Metric`. Given this PR is merged, shall I raise an issue for this?

Reply:
The check would be in the LighterSystem. But if we're doing it for metrics, we should then type-check everything:
https://github.com/project-lighter/lighter/blob/08cb305e965b9434411c59b752db40c454a5dc40/lighter/system.py#L20

That becomes quite a hassle, I guess this is when I wish there was runtime type checking in Python.

Reply:
@surajpaib wdy think?